python/cuml/manifold/umap.pyx

```diff
@@ -723,24 +780,41 @@ class UMAP(Base):
         cdef handle_t * handle_ = \
             <handle_t*> <size_t> self.handle.getHandle()
 
-        cdef uintptr_t orig_x_raw = self._X_m.ptr
         cdef uintptr_t embed_ptr = self._embedding_.ptr
 
         cdef UMAPParams* umap_params = \
             <UMAPParams*> <size_t> UMAP._build_umap_params(self)
 
-        transform(handle_[0],
-                  <float*>x_ptr,
-                  <int> X_m.shape[0],
-                  <int> X_m.shape[1],
-                  <int64_t*> knn_indices_raw,
-                  <float*> knn_dists_raw,
-                  <float*>orig_x_raw,
-                  <int> self.n_rows,
-                  <float*> embed_ptr,
-                  <int> self.n_rows,
-                  <UMAPParams*> umap_params,
-                  <float*> xformed_ptr)
+        if self._sparse_fit:
+            transform_sparse(handle_[0],
+                             <int*><uintptr_t> X_m.indptr.ptr,
+                             <int*><uintptr_t> X_m.indices.ptr,
+                             <float*><uintptr_t> X_m.data.ptr,
+                             <size_t> X_m.nnz,
+                             <int> X_m.shape[0],
+                             <int> X_m.shape[1],
+                             <int*><uintptr_t> self._X_m.indptr.ptr,
+                             <int*><uintptr_t> self._X_m.indices.ptr,
+                             <float*><uintptr_t> self._X_m.data.ptr,
+                             <size_t> self._X_m.nnz,
+                             <int> self._X_m.shape[0],
+                             <float*> embed_ptr,
+                             <int> self.n_rows,
+                             <UMAPParams*> umap_params,
+                             <float*> xformed_ptr)
+        else:
+            transform(handle_[0],
+                      <float*><uintptr_t> X_m.ptr,
+                      <int> X_m.shape[0],
+                      <int> X_m.shape[1],
+                      <int64_t*> knn_indices_raw,
+                      <float*> knn_dists_raw,
+                      <float*><uintptr_t>self._X_m.ptr,
+                      <int> self.n_rows,
+                      <float*> embed_ptr,
+                      <int> self.n_rows,
+                      <UMAPParams*> umap_params,
+                      <float*> xformed_ptr)
```

Comment:
These are really long calls. Maybe it would be nice to use the structs inheriting from the `manifold_inputs_t` struct in the Cython layer to have a clearer/simpler code.

Reply:
I've been thinking about this one throughout the day and I'm not sure exposing the `manifold_inputs_t` implementations through Cython is going to add a lot of complexity to that layer, mostly because the `manifold_inputs_t` has a class template, which means we will also need to deal with its instantiations. I do really like the idea of maintaining a clean API in general but I think the current C++ public API isn't too bad. 

On the other hand, I think TSNE could benefit from a `TSNEParams` class.